src/main/java/org/spongepowered/api/data/DataManipulator.java

```diff
@@ -0,0 +1,448 @@
+ * such that the {@link DataManipulator} is always returned.</p>
+ */
+public interface DataManipulator extends ValueContainer {
+
+    @TransformWith
+    @Override
+    DataManipulator copy();
+
+    Mutable asMutableCopy();
+
+    /**
+     * Gets a {@link Mutable} copy of this
+     * {@link DataManipulator} such that all backed
+     * {@link Value}s are copied into their {@link org.spongepowered.api.data.value.Value.Mutable}
+     * counterparts. Any changes to this {@link DataManipulator} will
+     * NOT be reflected on the returned {@link Mutable} and vice versa.
+     *
+     * <p>The cases where this {@link DataManipulator} is already a
+     * {@link Mutable}, the same manipulator object is returned.</p>
+     *
+     * @return This {@link DataManipulator}'s data copied into a
+     *     {@link Mutable}
+     */
+    Mutable asMutable();
+
+    /**
+     * Gets an {@link Immutable} copy of this
+     * {@link DataManipulator} such that all backed {@link org.spongepowered.api.data.value.Value.Mutable}s are copied
+     * into {@link org.spongepowered.api.data.value.Value.Immutable} counterparts. Any changes to this
+     * {@link DataManipulator} will NOT be reflected on the returned
+     * {@link Immutable} and vice versa.
+     *
+     * @return This {@link DataManipulator}'s data copied into a
+     *     {@link Immutable}
+     */
+    Immutable asImmutable();
+
+    interface Immutable extends DataManipulator {
+
+        @SuppressWarnings("unchecked")
+        @Override
+        default Immutable copy() {
+            return this;
+        }
+
+        @Override
+        default Immutable asImmutable() {
+            return this;
+        }
+
+        Mutable asMutable();
+
+        /**
+         * Creates a {@link Immutable} view directly based on the
+         * {@link Value}s. No unnecessary copies of the {@link Value}s
+         * will be created.
+         *
+         * @param values The values
+         * @return The immutable data manipulator view
+         */
+        static Immutable viewOf(Iterable<Value<?>> values) {
+            return Sponge.getRegistry().requireFactory(Factory.class).viewOf(values);
+        }
+
+        static Immutable viewOf(ValueContainer valueContainer) {
+            return Sponge.getRegistry().requireFactory(Factory.class).viewOf(valueContainer);
+        }
+
+        /**
+         * Gets a empty {@link Immutable}.
+         *
+         * @return The empty immutable data manipulator
+         */
+        static Immutable empty() {
+            return Sponge.getRegistry().requireFactory(Factory.class).of();
+        }
+
+        /**
+         * Gets a {@link Immutable} with the given values.
+         *
+         * @param values The values
+         * @return The immutable data manipulator
+         */
+        static Immutable of(Iterable<? extends Value<?>> values) {
+            return Sponge.getRegistry().requireFactory(Factory.class).of(values);
+        }
```

Comment:
The factory methods should be moved to `DataManipulator` and be renamed. E.g. `DataManipulator.immutableOf` instead of `DataManipulator.Immutable.of`. IMO it feels cleaner.

Reply:
`ofImmutable` for naming.

Reply:
:+1: for `immutableOf`.